cpp/src/slice2cs/CsUtil.cpp

```diff
@@ -1103,223 +1105,123 @@ Slice::CsGenerator::sequenceUnmarshalCode(const SequencePtr& seq, const string&
+                                                   const SequencePtr& seq,
+                                                   const string& scope,
+                                                   const string& param,
+                                                   int tag,
+                                                   const string& stream)
+{
+    const TypePtr type = seq->type();
+    const bool isArray = !seq->hasMetaDataWithPrefix("cs:generic:");
+    const string length = isArray ? param + ".Length" : param + ".Count";
+    BuiltinPtr builtin = BuiltinPtr::dynamicCast(type);
+    StructPtr st = StructPtr::dynamicCast(type);
+
+    if(isVariableLength(type))
     {
-        out << nl << "if(" << param << " != null && " << stream << ".WriteOptional(" << tag << ", "
+        out << nl << "if (" << param << " != null && " << stream << ".WriteOptional(" << tag << ", "
             << getTagFormat(seq, scope) << "))";
         out << sb;
         out << nl << "var pos = " << stream << ".StartSize();";
         writeMarshalCode(out, seq, scope, param, stream);
         out << nl << stream << ".EndSize(pos);";
         out << eb;
     }
-    else
+    else if(st)
     {
-        out << nl << "if(" << stream << ".ReadOptional(" << tag << ", " << getTagFormat(seq, scope) << "))";
+        out << nl << "if (" << param << " != null && " << stream << ".WriteOptional(" << tag << ", "
+            << getTagFormat(seq, scope) << "))";
         out << sb;
-        out << nl << stream << ".Skip(4);";
-        string tmp = "tmpVal";
-        out << nl << seqS << ' ' << tmp << ';';
-        writeUnmarshalCode(out, seq, scope, tmp, stream);
-        if(isArray)
+        if(st->minWireSize() > 1)
         {
-            out << nl << param << " = " << tmp << ";";
-        }
-        else
-        {
-            out << nl << param << " = new " << seqS << "(" << tmp << ");";
+            out << nl << stream << ".WriteSize(" << length << " * " << st->minWireSize() << " + (" << length
+                << " > 254 ? 5 : 1));";
         }
+        writeMarshalCode(out, seq, scope, param, stream);
         out << eb;
-        out << nl << "else";
+    }
+    else if(seq->hasMetaDataWithPrefix("cs:serializable:"))
+    {
+        out << nl << "if (" << param << " != null && " << stream << ".WriteOptional(" << tag
+            << ", " << getUnqualified("Ice.OptionalFormat", scope) << ".VSize))";
         out << sb;
-        out << nl << param << " = null;";
+        out << nl << stream << ".WriteSerializable(" << param << ");";
         out << eb;
     }
+    else if(isArray)
+    {
+        out << nl << stream << ".Write" << builtinSuffixTable[builtin->kind()] << "Seq(" << tag << ", "
+            << param << ");";
+    }
+    else
+    {
+        out << nl << "if (" << param << " != null)";
+        out << sb;
+        out << nl << stream << ".Write" << builtinSuffixTable[builtin->kind()] << "Seq(" << tag << ", " << param
```

Comment:
This "param << param <<" does not look correct. And the WriteXxxSeq on OutputStream perform the null check.

Reply:
This check was not required, remove it.